src/transformers/models/modernbert_decoder/modular_modernbert_decoder.py

```diff
@@ -0,0 +1,1305 @@
+
+            # Reset mask and apply combined constraints
+            causal_mask = torch.full(
+                (seq_len, kv_seq_len),
+                torch.finfo(attn_weights.dtype).min,
+                device=attn_weights.device,
+                dtype=attn_weights.dtype,
+            )
+            causal_mask = causal_mask.masked_fill(combined_pattern, 0)
+
+        attn_weights = attn_weights + causal_mask
+
+    # Apply attention mask (for padding) if provided
+    if attention_mask is not None:
+        # Check if the attention mask has the correct kv dimension
+        if attention_mask.dim() == 4 and attention_mask.shape[-1] != kv_seq_len:
+            # The mask was pre-computed but doesn't match our kv sequence length
+            # This happens when we have past key values that extend the sequence
+            current_mask_kv_len = attention_mask.shape[-1]
+
+            if current_mask_kv_len < kv_seq_len:
+                # Need to extend the mask to cover past keys
+                # Past keys should be allowed (not masked), so we pad with zeros
+                past_length = kv_seq_len - current_mask_kv_len
+
+                # Create padding for past positions (zeros = allow attention)
+                past_padding = torch.zeros(
+                    (attention_mask.shape[0], attention_mask.shape[1], attention_mask.shape[2], past_length),
+                    dtype=attention_mask.dtype,
+                    device=attention_mask.device,
+                )
+
+                # Concatenate past padding with current mask
+                attention_mask = torch.cat([past_padding, attention_mask], dim=-1)
+
+        attn_weights = attn_weights + attention_mask
+
+    # upcast attention to fp32
+    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+    attn_weights = nn.functional.dropout(attn_weights, p=dropout, training=module.training)
+    attn_output = torch.matmul(attn_weights, value)
+
+    # Return in 4D format: [batch_size, seq_len, num_heads, head_dim] so existing reshape logic works
+    attn_output = attn_output.transpose(1, 2).contiguous()
+
+    return attn_output, attn_weights
+
+
+class ModernBertDecoderAttention(nn.Module):
+    """Performs causal multi-headed self attention for ModernBERT decoder.
+
+    This module implements causal (unidirectional) attention, suitable for language modeling tasks.
+    It supports both local attention (sliding window) and global attention patterns.
+    """
+
+    def __init__(self, config: ModernBertDecoderConfig, layer_id: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        self.layer_id = layer_id
+
+        if config._attn_implementation not in ["eager", "flash_attention_2"]:
+            raise ValueError(
+                f"ModernBertDecoder only supports 'eager' and 'flash_attention_2' attention implementations. "
+                f"Got: {config._attn_implementation}"
+            )
```

Comment:
_supports_flex_attention and _supports_sdpa are here for this for the PreTrainedModel